harness/determined/common/experimental/checkpoint/_checkpoint.py

```diff
@@ -323,11 +312,15 @@ def remove_metadata(self, keys: List[str]) -> None:
             keys (List[string]): Top-level keys to remove from the checkpoint metadata.
         """
 
+        updated_metadata = dict(self.metadata)
         for key in keys:
-            if key in self.metadata:
-                del self.metadata[key]
+            if key in updated_metadata:
+                del updated_metadata[key]
+
+        req = _metadata_update_request(self.uuid, updated_metadata)
+        bindings.post_PostCheckpointMetadata(self._session, body=req, checkpoint_uuid=self.uuid)
 
-        self._push_metadata()
+        self.metadata = updated_metadata
```

Comment:
should we use the response from the API call `bindings.post_PostCheckpointMetadata` to update the Checkpoint state instead? since this is to main which doesn't have reload() logic, it could mean that if another process updates the checkpoint metadata, this object would have an incorrect state.

Reply:
It's a good idea, and you're right to bring it up. I'm not actually sure what the best choice is, so I think I've got a preference for "do less work." Here's a justification:

From the conversation we had OOB, I think we'd both been expecting `add_metadata` to do something other than what it actually does. `add_metadata` _sets_ metadata on master to `self.metadata | metadata` and `remove_metadata` _sets_ it to `self_metadata - keys`. The deltas aren't passed through to master; the locally-computed result is.

So for `updated_metadata` to _not_ equal the result the `bindings.post_PostCheckpointMetadata` would be deeply weird, and not the result of a race condition.